### PR TITLE
Add file latest and file md5 latest to south API router

### PR DESCRIPTION
**Pull request by WangWinson**
To solve issue https://github.com/RackHD/on-http/issues/215
- Add file latest and file md5 latest router to south file router
- Add unit test

@RackHD/corecommitters @cgx027

### CODE REVIEW

**Comment by WangWinson** [comment]
test this please

**Comment by WangWinson** [comment]
test this please

File: lib/api/1.1/southbound/files.js [hunk 1]
```diff
@@ -71,6 +71,59 @@ function filesRouterFactory (
...
+    router.get('/files/:filename/latest', function(req, res) {
+        logger.debug("Received request for file by uuid", {
+            uuid: req.params.uuid
+        });
+
+        var query = req.params.filename;
+
+        return fileService.verify(query)
+        .then(function(metadata) {
+            return metadata[0].uuid;
```
Comment (iceiilin) [inline lib/api/1.1/southbound/files.js]: Change to metadata[metadata.length - 1] to return the latest file. Similar to https://github.com/RackHD/on-http/pull/222

File: lib/api/1.1/southbound/files.js [hunk 2]
```diff
@@ -124,6 +177,45 @@ function filesRouterFactory (
...
+     */
+    router.get('/files/md5/:filename/latest', function(req, res) {
+        logger.debug("Received md5sum request for file" +
+            req.params.filename + " from the global files collection.");
+
+        var query = req.params.filename;
+
+        return fileService.verify(query)
+        .then(function(metadata) {
+            res.json(200,  metadata[0].md5);
```
Comment (iceiilin) [inline lib/api/1.1/southbound/files.js]: Similar as above

File: lib/api/1.1/southbound/files.js [hunk 3]
```diff
@@ -71,6 +71,59 @@ function filesRouterFactory (
...
+        .catch(function(err) {
+            logger.error("Failure serving file request", { error: err });
+            if (err.name === '404') {
+                res.status(404).json({
+                    error: 'File not found.'
+                });
+            } else {
+                res.status(500).json({
+                    error: "Failed to serve file request."
+
```
Comment (yyscamper) [inline lib/api/1.1/southbound/files.js]: this empty line is not necessary.

**Comment by yyscamper** [comment]
Still :+1:

**Comment by WangWinson** [comment]
Thanks @iceiilin, just see your PR merged.

**Comment by iceiilin** [comment]
+1 after changing the index

**Comment by anhou** [comment]
@WangWinson suggest adding unittests for these two routers

**Comment by WangWinson** [comment]
@iceiilin @yyscamper comments addressed
Agree @anhou, though the file router missing UT originally, it's better to add UT. This PR could pending here to wait UT if no urgent require to merge. 
PT story https://www.pivotaltracker.com/projects/1492892/stories/117889937 created to track this effort.

**Comment by cgx027** [comment]
Looks good to me :+1:

**Comment by amymullins** [comment]
@WangWinson can we please include the unit tests with this fix?  It would also likely have a positive impact to the coveralls decreased coverage noted above. thanks!

**Comment by WangWinson** [comment]
@amymullins yes, I would like to. I would try best to complete my current story first.

**Comment by zyoung51** [comment]
@WangWinson These are the coverage/unittest rules Felix has posted for me in the past:
(1) If the new code totally doesn't have any unit-test code, then it is gating; 
(2) If there is a lot of new code and reasonable amount of unit-test, then it is OK for me as long as the coverage doesn't decrease a lot ( less than around 0.5% in my practice).

+1 after the PR is amended to meet these criteria, -1 otherwise.

**Comment by yyscamper** [comment]
@amymullins @zyoung51 : I had talked about  the unit-test with @WangWinson, since there is a history debt that this file doesn't have any unit-test before and considering the priority that this will impact multiple P1 issues that related with firmware update, so I'm acceptable to create another PR to compensate the unit-testing.

Anyway, I will be very appreciated if the unit-test can be ready in this PR.

**Comment by amymullins** [comment]
@yyscamper ok by me if you need/want to track in a separate story but IMO it should be of equal priority.  If we had these test cases covered, we may not have had multiple P1 issues to resolve. :)

**Comment by yyscamper** [comment]
@amymullins : I agree. 
What goes around comes around and You reap what you sow  :)

File: spec/lib/api/1.1/files-spec.js [hunk 4]
```diff
@@ -3,6 +3,360 @@
...
+    describe('GET /files/:uuid', function() {
+        before('create mock file stream', function() {
+            mockRdStream = new stream.Readable();
+            mockRdStream._read = function() {
+                this.push('this is mock file 1234');
+                this.push(null);
+            };
+        });
+
+        beforeEach('set up mocks', function() {
```
Comment (yyscamper) [inline spec/lib/api/1.1/files-spec.js]: setup?

**Comment by yyscamper** [comment]
:+1:

**Comment by iceiilin** [comment]
:+1:

**Pull request merged**